File: arrays/dynamic.py
class Dynamic_Array:
    def __init__(self):
        self._size = 0
        self._capacity = 1
        self._Array = [None] * self._capacity

    def size(self):
        return self._size

    def capacity(self):
        return self._capacity

    def is_empty(self):
        if self._size == 0:
            return True
        return False

    def at(self, i):
        if i < 0 or i >= self._capacity:
            raise IndexError("Out of range.")
        return self._Array[i]

    def push(self, item):
        if self._capacity == self._size:
            self._resize(2 * self._capacity)
        self._Array[self._size] = item
        self._size += 1

    def insert(self, index, item):
        if self._capacity == self._size:
            self._resize(2 * self._capacity)
        for i in range(self._size, index, -1):
            self._Array[i] = self._Array[i - 1]
        self._Array[index] = item
        self._size += 1

    def prepend(self, item):
        self.insert(0, item)

    def delete(self, i):
        if i < 0 or i > self._size:
            raise IndexError("Out of range.")
        for j in range(i, self._size - 1):
            self._Array[j] = self._Array[j + 1]
        self._size -= 1
    
    def find(self, item):
        try:
            return self._Array.index(item)
        except ValueError:
            return -1

    def _resize(self, capacity):
        NewArray = [None] * (capacity)
        for i in range(self._size):
            NewArray[i] = self._Array[i]
        self._Array = NewArray
        self._capacity = capacity

    def __str__(self):
        message = ""
        for i in range(self._size):
            message += str(self._Array[i])
            message += " "
        return message

File: arrays/dynamic.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

int CAPACITY;
void display(int *);
int size(int *);
bool is_empty(int *);
void at(int *, int);
int *push(int *, int);
int *insert(int *, int, int);
int *prepend(int *, int);
int pop(int *);
int *delete(int *, int);
int *rem(int *, int);

int main(void)
{
    int *a;
    printf("Enter no. of elements to store: ");
    scanf("%d", &CAPACITY);
    a = malloc((CAPACITY + 1) * sizeof(int));
    printf("Storing %d integers...\n", CAPACITY);
    for (int i = 0; i < CAPACITY; i++)
        *(a + i) = i;
    *(a + CAPACITY) = -999;

    if (is_empty(a))
        printf("\nEmpty\n");
    display(a);

    at(a, 1);
    a = push(a, 99);
    display(a);

    a = insert(a, 2, 99);
    display(a);

    a = prepend(a, 99);
    display(a);

    int p = pop(a);
    printf("\nPoped element: %d\n", p);
    display(a);

    a = delete (a, 5);
    display(a);

    a = rem(a, 99);
    display(a);


    return 0;
}

void display(int *a)
{
    int i = 0;
    printf("Size: %d | Capacity: %d\n", size(a), CAPACITY);
    printf("----Integers:----\n");
    while (*(a + i) != -999)
    {
        printf("%d: %d\n", i, *(a + i));
        i++;
    }
}
int size(int *a)
{
    int i = 0;
    while (*(a + i) != -999)
        i++;
    return i;
}
bool is_empty(int *a)
{
    if (size(a) == 0)
        return true;
    return false;
}
void at(int *a, int i)
{
    if (i >= CAPACITY)
        printf("\nIndex out of bound\n");
    else
        printf("\nReturning element at %d: %d\n", i, *(a + i));
}
int *push(int *a, int x)
{
    printf("\n+ + + + Pushing %d + + + +\n", x);
    CAPACITY += 1;
    a = realloc(a, sizeof(int) * CAPACITY + 1);
    *(a + CAPACITY - 1) = x;
    *(a + CAPACITY) = -999;
    return a;
}
int *insert(int *a, int index, int x)
{
    CAPACITY += 1;
    int i = CAPACITY;
    a = realloc(a, sizeof(int) * CAPACITY + 1);
    printf("\n* * * Inserting %d at %d * * *\n", x, index);
    while (i >= index)
    {
        *(a + i) = *(a + i - 1);
        i--;
    }
    *(a + index) = x;
    return a;
}
int *prepend(int *a, int x)
{
    CAPACITY += 1;
    int i = CAPACITY;
    a = realloc(a, sizeof(int) * CAPACITY + 1);
    printf("\n- - - - Prepending %d - - - -\n", x);
    while (i > 0)
    {
        *(a + i) = *(a + i - 1);
        i--;
    }
    *a = x;
    return a;
}
int pop(int *a)
{
    int p = *(a + size(a) - 1);
    CAPACITY--;
    a = realloc(a, sizeof(int) * CAPACITY + 1);
    *(a + CAPACITY) = -999;
    return p;
}
int *delete(int *a, int i)
{
    printf("\n- - - Deleting item at %d - - -\n", i);
    while (i < CAPACITY)
    {
        *(a + i) = *(a + i + 1);
        i++;
    }
    CAPACITY--;
    a = realloc(a, sizeof(int) * CAPACITY + 1);
    *(a + CAPACITY) = -999;
    return a;
}

int *rem(int *a, int x)
{
    for (int i = 0; i < CAPACITY; i++)
    {
        if (*(a + i) == x)
            a = delete (a, i);
    }
    return a;
}
